src/shared/configUtils.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { Workspace } from '@lwc/lwc-dev-server';
+import { CryptoUtils, SSLCertificateData } from '@salesforce/lwc-dev-mobile-core';
+import { Config, ConfigAggregator } from '@salesforce/core';
+import configMeta, { ConfigVars } from './../configMeta.js';
+
+export const LOCAL_DEV_SERVER_DEFAULT_PORT = 8081;
+export const LOCAL_DEV_SERVER_DEFAULT_WORKSPACE = Workspace.SfCli;
+
+export class ConfigUtils {
+  static #config: Config;
+  static #globalConfig: Config;
+
+  public static async getConfig(): Promise<Config> {
+    if (this.#config) {
+      return this.#config;
+    }
+    this.#config = await Config.create({ isGlobal: false });
+    Config.addAllowedProperties(configMeta);
+    return this.#config;
+  }
+
+  public static async getGlobalConfig(): Promise<Config> {
+    if (this.#globalConfig) {
+      return this.#globalConfig;
+    }
+    this.#globalConfig = await Config.create({ isGlobal: true });
+    Config.addAllowedProperties(configMeta);
+    return this.#globalConfig;
+  }
+
+  public static async getOrCreateIdentityToken(): Promise<string> {
+    let token = await this.getIdentityToken();
+    if (!token) {
+      token = CryptoUtils.generateIdentityToken();
+      await this.writeIdentityToken(token);
+    }
+    return token;
+  }
+
+  public static async getIdentityToken(): Promise<string | undefined> {
+    const config = await ConfigAggregator.create({ customConfigMeta: configMeta });
+    // Need to reload to make sure the values read are decrypted
+    await config.reload();
+    const identityToken = config.getPropertyValue(ConfigVars.LOCAL_WEB_SERVER_IDENTITY_TOKEN);
+
+    return identityToken as string;
+  }
+
+  public static async writeIdentityToken(token: string): Promise<void> {
+    const config = await this.getConfig();
+    config.set(ConfigVars.LOCAL_WEB_SERVER_IDENTITY_TOKEN, token);
+    await config.write();
+  }
+
+  public static async getCertData(): Promise<SSLCertificateData | undefined> {
+    const config = await this.getGlobalConfig();
+    const data = config.get(ConfigVars.LOCAL_DEV_SERVER_HTTPS_CERT_DATA) as SSLCertificateData;
+    // convert base64 to binary DER data
+    if (data?.derCertificate) {
+      data.derCertificate = Buffer.from(data.derCertificate, 'base64').toString('binary');
```

Comment:
This is a dicey way to manage binary data, keeping it in a `string`. You can see from Node.js's `Buffer` [docs](https://nodejs.org/docs/latest-v20.x/api/buffer.html#buffers-and-character-encodings), the following relevant verbiage:

> The following legacy character encodings are also supported:
> ...
> `'binary'`: Alias for `'latin1'`. The name of this encoding can be very misleading, as all of the encodings listed here convert between strings and binary data. For converting between strings and `Buffer`s, typically `'utf8'` is the right choice.

The proper way to manage pure binary data in memory, with Node, is in a `Buffer`. I propose that `SSLCertificateData.derCertificate` should be a `Buffer`, not a `string`. It actually simplifies operations on that data:

- No converting the data member into a `Buffer`.
- No messing with encoding.
- No need to further configure file operations, because file reads and writes take a `Buffer` as their outputs and inputs, respectively.

Reply:
Ah, following that back through `lwc-dev-mobile-core`, I see it goes all the way back to node-forge and the fact that their representation of a binary (`Bytes`) type is just an alias to `string`. I don't know why they made that choice, but...¯\\_(ツ)_/¯.

I'd still argue that it's cleaner, more concise, and more representative of the nature of the data, to have it as a `Buffer`. But that would require converting it in `lwc-dev-mobile-core`, via `Buffer.from()` where it's [created](https://github.com/forcedotcom/lwc-dev-mobile-core/blob/d974aefb7c85e3b0687167b5f146b362b62cf202/src/common/CryptoUtils.ts#L124). What are your thoughts?

Reply:
Reading through some of https://github.com/digitalbazaar/forge/issues/146, I'm guessing the design decision was made around flexibility of the code (e.g. use outside of Node.js). But we're standardized on Node.js, so I'd be partial to "converting" to the idioms of our platform.

Reply:
I see what you mean. As you've discovered already, this is b/c of how node-forge works. Specifically [this line](https://github.com/maliroteh-sf/lwc-dev-mobile-core/blob/main/src/common/CryptoUtils.ts#L124) is where we generate the DER data which node-forge returns back as string (`getBytes()` returns `Bytes` and node-forge has `type Bytes = string;`)

I do have a change coming up in `lwc-dev-mobile-core` and can go ahead and make some changes for this too. So your proposal is to essentially do something along the lines of
```
const derCert = forge.asn1.toDer(forge.pki.certificateToAsn1(cert)).getBytes();
const data = Buffer.from(derCert, 'binary');

return {
    derCertificate: data,
    pemCertificate: pemCert,
    pemPrivateKey: privateKey,
    pemPublicKey: publicKey
};
```
Then here in `configUtils` we would convert the buffer to base64 when saving to global config, and conversely convert back from base64 to buffer when reading from global config?

Reply:
💯 Yeah that's exactly it. And then working with `Buffer`s on the file operations side is more concise as well.